storage-provider/server/src/pipeline/add_piece.rs

```diff
@@ -0,0 +1,248 @@
+            if old_cancellation_sender.send(()).is_err() {
+                tracing::error!("Failed to send cancellation value");
+            }
+
+            tracing::debug!(%sector_number, "Existing deadline & task have been replaced: new_deadline = {}", deadline);
+        }
+        Some((old_deadline, _)) => {
+            tracing::debug!(%sector_number, "Existing deadline is newer than the new one: old_deadline = {}, new_deadline = {}", old_deadline, deadline);
+        }
+        None => {
+            tracing::debug!(%sector_number, "No deadline existing existed, creating it.");
+            let (cancellation_sender, cancellation_receiver) = oneshot::channel();
+            let state_for_task = state.clone();
+            schedule_send_pre_commit(
+                state_for_task,
+                tracker,
+                sector_number,
+                when,
+                cancellation_receiver,
+            );
+            scheduled_pre_commits.insert(sector_number, (deadline, cancellation_sender));
+        }
+    }
+}
+
+/// *Unconditionally* schedule a pre commit.
+fn schedule_send_pre_commit(
+    state: Arc<PipelineState>,
+    tracker: TaskTracker,
+    sector_number: SectorNumber,
+    when: Duration,
+    cancellation_receiver: oneshot::Receiver<()>,
+) {
+    let span = tracing::info_span!("add_piece");
+    tracing::info!(
+        // Since the span is moved for the instrument call, we can't span.enter()
+        // doing a span.in_scope(|| ...) is also a bit overkill for a log
+        parent: &span,
+        "Launching task to pre-commit in {} seconds",
+        when.as_secs()
+    );
+    // We don't care for the returned JoinHandle, but that's ok because the TaskTracker has it!
+    let _ = tracker.spawn(
+        async move {
+            match tokio::time::timeout(when, cancellation_receiver).await {
+                Ok(Ok(())) => {
+                    tracing::debug!(%sector_number, "Received cancellation signal, not sending message.");
+                    return;
+                },
+                Ok(Err(err)) => {
+                    tracing::error!(%sector_number, "Failed to receive message with error (will return): {err}");
+                    return;
+                },
+                Err(_elapsed) => tracing::debug!(%sector_number, "No cancelation signal was received"),
+            }
+
+            tracing::info!(%sector_number, "Awoken from sleep, submitting pre-commit task!");
+            match state.send_pre_commit(sector_number) {
+                Ok(()) => tracing::info!(%sector_number, "Successfully submitted task!"),
+                // Not sure what we can do if the task fails, maybe try to re-submit?
+                // Not even sure when this can happen asides from an OOM since we're using an unbounded channel
+                Err(err) => tracing::error!(%sector_number, "Failed to submit task: {}", err),
+            }
+        }
+        .instrument(span),
```

Comment:
```suggestion
    let _ = tracker.spawn(
        async move {
            tokio::select! {
                _ = tokio::time::sleep(when) => {
                    tracing::debug!(%sector_number, "No cancellation signal was received");
                    tracing::info!(%sector_number, "Awoken from sleep, submitting pre-commit task!");
                    match state.send_pre_commit(sector_number) {
                        Ok(()) => tracing::info!(%sector_number, "Successfully submitted task!"),
                        Err(err) => tracing::error!(%sector_number, "Failed to submit task: {}", err),
                    }
                }
                _ = cancellation_receiver => {
                    tracing::debug!(%sector_number, "Received cancellation signal, not sending message.");
                }
            }
        }
        .instrument(span),
```

Have you considered doing it like this?